src/partition/repart.c

```diff
@@ -3280,12 +3282,70 @@ static int do_make_directories(Partition *p, const char *root) {
         return 0;
 }
 
-static int partition_populate(Partition *p, const char *node) {
+static int partition_populate_directory(Partition *p, char **ret_root, char **ret_tmp_root) {
+        _cleanup_free_ char *root = NULL;
+        int r;
+
+        assert(ret_root);
+        assert(ret_tmp_root);
+
+        /* When generating squashfs, we need the directory root to be available when we call squashfs.
+         * Because we might have multiple source trees, we build a temporary source tree beforehand where we
+         * merge all our inputs. We then use this merged source tree to create the squashfs filesystem. */
+
+        if (!streq(p->format, "squashfs")) {
+                *ret_root = NULL;
+                *ret_tmp_root = NULL;
+                return 0;
+        }
+
+        /* If we only have a single directory that's meant to become the root directory of the filesystem,
+         * we can shortcut this function and just use that directory as the root directory instead. If we
+         * allocate a temporary directory, it's stored in `ret_tmp_root` to indicate it should be removed.
+         * Otherwise, we return the directory to use in `root` to indicate it should not be removed. */
+
+        if (strv_length(p->copy_files) == 2 && strv_length(p->make_directories) == 0 && streq(p->copy_files[1], "/")) {
```

Comment:
what happens if you want to create an empty squasfs?

Reply:
We create an empty directory and return that? I wonder if we should explicitly disallow empty squashfs, since there's not really a point to an empty read-only filesystem I think?